scripts/nmt/encoder_decoder.py

```diff
@@ -584,6 +595,599 @@ def hybrid_forward(self, F, step_input, states):  #pylint: disable=arguments-dif
+    # Apply the cosine to even columns and sin to odds.
+    position_enc[:, 0::2] = np.sin(position_enc[:, 0::2])  # dim 2i
+    position_enc[:, 1::2] = np.cos(position_enc[:, 1::2])  # dim 2i+1
+    return position_enc
+
+
+class PositionwiseFFN(HybridBlock):
+    """Structure of the Positionwise Feed-Forward Neural Network.
+
+    Parameters
+    ----------
+    units : int
+    hidden_size : int
+        number of units in the hidden layer of position-wise feed-forward networks
+    dropout : float
+    use_residual : bool
+    weight_initializer : str or Initializer
+        Initializer for the input weights matrix, used for the linear
+        transformation of the inputs.
+    bias_initializer : str or Initializer
+        Initializer for the bias vector.
+    activation : str, default 'relu'
+    	Activation function
+    prefix : str, default 'rnn_'
+        Prefix for name of `Block`s
+        (and name of weight if params is `None`).
+    params : Parameter or None
+        Container for weight sharing between cells.
+        Created if `None`.
+    """
+    def __init__(self, units=512, hidden_size=2048, dropout=0.0, use_residual=True,
+                 weight_initializer=None, bias_initializer='zeros', activation='relu',
+                 prefix=None, params=None):
+        super(PositionwiseFFN, self).__init__(prefix=prefix, params=params)
+        self._hidden_size = hidden_size
+        self._units = units
+        self._use_residual = use_residual
+        self.ffn_1 = nn.Dense(units=hidden_size, flatten=False, activation=activation,
+                              weight_initializer=weight_initializer, bias_initializer=bias_initializer,
+                              prefix='ffn_1_')
+        self.ffn_2 = nn.Dense(units=units, flatten=False,
+                              weight_initializer=weight_initializer, bias_initializer=bias_initializer,
+                              prefix='ffn_2_')
+        self.dropout_layer = nn.Dropout(dropout)
+        self.layer_norm = nn.LayerNorm()
+
+    def hybrid_forward(self, F, inputs):
+        outputs = self.ffn_1(inputs)
+        outputs = self.ffn_2(outputs)
+        outputs = self.dropout_layer(outputs)
+        if self._use_residual:
+            outputs = outputs + inputs
+        outputs = self.layer_norm(outputs)
+        return outputs
+
+
+class TransformerEncoderCell(HybridBlock):
+    """Structure of the Transformer Encoder Cell.
+
+    Parameters
+    ----------
+    attention_cell : AttentionCell or str, default 'multi_head'
+        Arguments of the attention cell.
+    units : int
+    hidden_size : int
+        number of units in the hidden layer of position-wise feed-forward networks
+    num_heads : int
+        Number of heads in multi-head attention
+    scaled : bool
+        Whether to scale the softmax input by the sqrt of the input dimension in multi-head attention
+    dropout : float
+    use_residual : bool
```

Comment:
nit: missing docs. maybe explain where the dropout and residual connection is used?